streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -0,0 +1,243 @@
+import org.apache.kafka.streams.processor.internals.RecordContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.SessionStore;
+import org.apache.kafka.streams.state.StateSerdes;
+
+import java.util.List;
+import java.util.NoSuchElementException;
+
+
+class CachingSessionStore<K, AGG>  implements SessionStore<K, AGG>, CachedStateStore<Windowed<K>, AGG> {
+
+    private final SegmentedBytesStore bytesStore;
+    private final RocksDBSessionStore.SessionKeySchema keySchema;
+    private Serde<K> keySerde;
+    private final Serde<AGG> aggSerde;
+    private InternalProcessorContext context;
+    private String name;
+    private StateSerdes<Windowed<K>, AGG> serdes;
+    private ThreadCache cache;
+    private CacheFlushListener<Windowed<K>, AGG> flushListener;
+
+    CachingSessionStore(final SegmentedBytesStore bytesStore,
+                        final Serde<K> keySerde,
+                        final Serde<AGG> aggSerde) {
+        this.bytesStore = bytesStore;
+        this.keySerde = keySerde;
+        this.aggSerde = aggSerde;
+        this.keySchema = new RocksDBSessionStore.SessionKeySchema();
+    }
+
+    public KeyValueIterator<Windowed<K>, AGG> findSessionsToMerge(final K key,
+                                                                  final long earliestSessionEndTime,
+                                                                  final long latestSessionStartTime) {
+        validateStoreOpen();
+        final Bytes binarySessionId = Bytes.wrap(keySerde.serializer().serialize(name, key));
+        final ThreadCache.MemoryLRUCacheBytesIterator cacheIterator = cache.range(name,
+                                                                                  keySchema.lowerRange(binarySessionId,
+                                                                                                       earliestSessionEndTime).get(),
+                                                                                  keySchema.upperRange(binarySessionId, latestSessionStartTime).get());
+        final KeyValueIterator<Bytes, byte[]> storeIterator = bytesStore.fetch(binarySessionId, earliestSessionEndTime, latestSessionStartTime);
+        final HasNextCondition<Bytes> hasNextCondition = keySchema.hasNextCondition(binarySessionId,
+                                                                                    earliestSessionEndTime,
+                                                                                    latestSessionStartTime);
+        final PeekingKeyValueIterator<Bytes, LRUCacheEntry> filteredCacheIterator = new FilteredCacheIterator(cacheIterator, hasNextCondition);
+        return new MergedSortedCacheKeyValueStoreIterator<>(filteredCacheIterator, storeIterator, serdes);
+    }
+
+
+    public void remove(final Windowed<K> sessionKey) {
+        validateStoreOpen();
+        put(sessionKey, null);
```

Comment:
`delete` on `CachingKeyValueStore` is not used anywhere except for tests - it probably should be removed. If we delete from a store we always put key:null.
I think there was another reason, too. But i can't remember what it was. It has been quite some time since this was written